### PR TITLE
jsonnet: Ingester automated downscale v2

### PR DESCRIPTION
**Pull request by pstibrany**
#### What this PR does

This PR adds support for new way of downscaling ingesters, using rollout-operator's resource-mirroring feature and read-only mode of ingesters. (see also https://github.com/grafana/mimir/pull/9132 and https://github.com/grafana/mimir/pull/9208). This new downscaling mode can be enabled by using `ingester_automated_downscale_v2_enabled` config option. This is mutually exclusive with both `ingester_automated_downscale_enabled` (previous downscale mode) and `ingest_storage_ingester_autoscaling_enabled` (autoscaling for ingest-storage).

#### Checklist

- [x] Tests updated.
- [na] Documentation added.
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [na] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

File: operations/mimir/ingester-automated-downscale-v2.libsonnet [hunk 1]
```diff
@@ -0,0 +1,99 @@
+{
+  // Allow ingesters to be safely downscaled using the "prepare-downscale" and "resource-mirroring" features
+  // of the rollout-operator. Only available when using multi-zone ingesters. This feature can be enabled
+  // via the ingester_automated_downscale_enabled flag.
```
Comment (pr00se) [inline operations/mimir/ingester-automated-downscale-v2.libsonnet:4]:
```suggestion
  // via the ingester_automated_downscale_v2_enabled flag.
```

File: operations/mimir/ingester-automated-downscale-v2.libsonnet [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+  // After this time is reached, rollout-operator will finally adjust number of replicas in ingester StatefulSets
+  // (for all zones).
+  //
+  // The number of replicas is still controlled by $._config.multi_zone_ingester_replicas until a reasonable metric to scale it
+  // automatically is determined and tested.
+  _config+: {
+    ingester_automated_downscale_v2_enabled: false,
+
+    // Allow to selectively enable it on a per-zone basis. Note that zone-a must be enabled first.
+    ingester_automated_downscale_v2_zone_a_enabled: $._config.ingester_automated_downscale_v2_enabled,
```
Comment (pr00se) [inline operations/mimir/ingester-automated-downscale-v2.libsonnet:24]: Do we want to keep this per-zone setting, or simplify it down to a single flag?

File: operations/mimir/ingester-automated-downscale-v2.libsonnet [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+                                                    $._config.ingester_automated_downscale_v2_zone_c_enabled,
+
+  // Validate the configuration.
+  assert !$._config.ingester_automated_downscale_v2_enabled || $._config.multi_zone_ingester_enabled : 'ingester downscaling requires multi_zone_ingester_enabled in namespace %s' % $._config.namespace,
+  assert !$._config.ingester_automated_downscale_v2_enabled || $._config.multi_zone_ingester_replicas > 0 : 'ingester downscaling requires multi_zone_ingester_replicas > 0 in namespace %s' % $._config.namespace,
+  assert !$._config.ingester_automated_downscale_v2_enabled || !$._config.ingester_automated_downscale_enabled : 'ingester_automated_downscale_enabled_v2 and ingester_automated_downscale_enabled are mutually exclusive in namespace %s' % $._config.namespace,
+  assert !$._config.ingester_automated_downscale_v2_zone_b_enabled || $._config.ingester_automated_downscale_v2_zone_a_enabled : 'ingester_automated_downscale_v2_zone_b_enabled enabled, but zone-a is not. ingester_automated_downscale_v2_zone_a_enabled must be enabled first in in namespace %s' % $._config.namespace,
```
Comment (pr00se) [inline operations/mimir/ingester-automated-downscale-v2.libsonnet:48]:
```suggestion
  assert !$._config.ingester_automated_downscale_v2_zone_b_enabled || $._config.ingester_automated_downscale_v2_zone_a_enabled : 'ingester_automated_downscale_v2_zone_b_enabled enabled, but zone-a is not. ingester_automated_downscale_v2_zone_a_enabled must be enabled first in zone-a in namespace %s' % $._config.namespace,
```

File: operations/mimir/ingester-automated-downscale-v2.libsonnet [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+  assert !$._config.ingester_automated_downscale_v2_zone_c_enabled || $._config.ingester_automated_downscale_v2_zone_a_enabled : 'ingester_automated_downscale_v2_zone_c_enabled enabled, but zone-a is not. ingester_automated_downscale_v2_zone_a_enabled must be enabled first in in namespace %s' % $._config.namespace,
```
Comment (pr00se) [inline operations/mimir/ingester-automated-downscale-v2.libsonnet:49]:
```suggestion
  assert !$._config.ingester_automated_downscale_v2_zone_c_enabled || $._config.ingester_automated_downscale_v2_zone_a_enabled : 'ingester_automated_downscale_v2_zone_c_enabled enabled, but zone-a is not. ingester_automated_downscale_v2_zone_a_enabled must be enabled first in zone-a in namespace %s' % $._config.namespace,
```

File: operations/mimir/ingest-storage-ingester-autoscaling.libsonnet [hunk 5]
```diff
@@ -29,6 +29,8 @@
...
   assert !$._config.ingest_storage_ingester_autoscaling_ingester_annotations_enabled || $._config.multi_zone_ingester_replicas <= 0 : 'partitions ingester autoscaling and multi_zone_ingester_replicas are mutually exclusive in namespace %s' % $._config.namespace,
   assert !$._config.ingest_storage_ingester_autoscaling_ingester_annotations_enabled || !$._config.ingester_automated_downscale_enabled : 'partitions ingester autoscaling and ingester automated downscale config are mutually exclusive in namespace %s' % $._config.namespace,
+  // We check ingester_automated_downscale_v2_zone_a_enabled here, because downscale in zone-b and zone-c can only be enabled after enabling zone-a.
+  assert !$._config.ingest_storage_ingester_autoscaling_ingester_annotations_enabled || !$._config.ingester_automated_downscale_v2_zone_a_enabled : 'partitions ingester autoscaling and ingester automated downscale v2 config are mutually exclusive in namespace %s' % $._config.namespace,
```
Comment (pr00se) [inline operations/mimir/ingest-storage-ingester-autoscaling.libsonnet:33]: Should we add a similar check to ingester-automated-downscale.libsonnet?

**Review by pr00se — APPROVED** [review]
Thank you!

File: operations/mimir/ingest-storage-ingester-autoscaling.libsonnet (hunk 5 above)
Comment (pstibrany) [reply operations/mimir/ingest-storage-ingester-autoscaling.libsonnet:33]: I don't quite follow. Doesn't this check cover it already?

File: operations/mimir/ingester-automated-downscale-v2.libsonnet (hunk 2 above)
Comment (pstibrany) [reply operations/mimir/ingester-automated-downscale-v2.libsonnet:24]: Let's simplify.

**Pull request merged**